// angular-bootstrap-datetimepicker/esm5/core/dl-date-time-core.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { DlDateAdapter } from './dl-date-adapter';
import { DlDateAdapterMoment } from './dl-date-adapter-moment';
import { DlDateAdapterNative } from './dl-date-adapter-native';
import { DlDateAdapterNumber } from './dl-date-adapter-number';
import { DlDateAdapterString } from './dl-date-adapter-string';
import { DL_DATE_TIME_DISPLAY_FORMAT, DL_DATE_TIME_DISPLAY_FORMAT_DEFAULT, DL_DATE_TIME_INPUT_FORMATS, DL_DATE_TIME_INPUT_FORMATS_DEFAULT, DL_DATE_TIME_MODEL_FORMAT, DL_DATE_TIME_MODEL_FORMAT_DEFAULT } from './dl-date-time-string-format';
var ɵ0 = DL_DATE_TIME_DISPLAY_FORMAT_DEFAULT, ɵ1 = DL_DATE_TIME_INPUT_FORMATS_DEFAULT, ɵ2 = DL_DATE_TIME_MODEL_FORMAT_DEFAULT;
/**
 * Import this module to supply your own `DateAdapter` provider.
 * \@internal
 *
 */
var DlDateTimeCoreModule = /** @class */ (function () {
    function DlDateTimeCoreModule() {
    }
    DlDateTimeCoreModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        { provide: DL_DATE_TIME_DISPLAY_FORMAT, useValue: ɵ0 },
                        { provide: DL_DATE_TIME_INPUT_FORMATS, useValue: ɵ1 },
                        { provide: DL_DATE_TIME_MODEL_FORMAT, useValue: ɵ2 }
                    ]
                },] }
    ];
    return DlDateTimeCoreModule;
}());
export { DlDateTimeCoreModule };
/**
 * Import this module to store `milliseconds` in the model.
 * \@internal
 */
var DlDateTimeNumberModule = /** @class */ (function () {
    function DlDateTimeNumberModule() {
    }
    DlDateTimeNumberModule.decorators = [
        { type: NgModule, args: [{
                    imports: [DlDateTimeCoreModule],
                    providers: [
                        { provide: DlDateAdapter, useClass: DlDateAdapterNumber }
                    ],
                    exports: [DlDateTimeCoreModule]
                },] }
    ];
    return DlDateTimeNumberModule;
}());
export { DlDateTimeNumberModule };
/**
 * Import this module to store a native JavaScript `Date` in the model.
 * \@internal
 */
var DlDateTimeDateModule = /** @class */ (function () {
    function DlDateTimeDateModule() {
    }
    DlDateTimeDateModule.decorators = [
        { type: NgModule, args: [{
                    imports: [DlDateTimeCoreModule],
                    providers: [
                        { provide: DlDateAdapter, useClass: DlDateAdapterNative }
                    ],
                },] }
    ];
    return DlDateTimeDateModule;
}());
export { DlDateTimeDateModule };
/**
 * Import this module to store a `moment` in the model.
 * \@internal
 */
var DlDateTimeMomentModule = /** @class */ (function () {
    function DlDateTimeMomentModule() {
    }
    DlDateTimeMomentModule.decorators = [
        { type: NgModule, args: [{
                    imports: [DlDateTimeCoreModule],
                    providers: [
                        { provide: DlDateAdapter, useClass: DlDateAdapterMoment }
                    ],
                },] }
    ];
    return DlDateTimeMomentModule;
}());
export { DlDateTimeMomentModule };
var ɵ3 = DL_DATE_TIME_INPUT_FORMATS_DEFAULT, ɵ4 = DL_DATE_TIME_DISPLAY_FORMAT_DEFAULT;
/**
 * Import this module to store a `string` in the model.
 * \@internal
 */
var DlDateTimeStringModule = /** @class */ (function () {
    function DlDateTimeStringModule() {
    }
    DlDateTimeStringModule.decorators = [
        { type: NgModule, args: [{
                    imports: [DlDateTimeCoreModule],
                    providers: [
                        { provide: DL_DATE_TIME_INPUT_FORMATS, useValue: ɵ3 },
                        { provide: DL_DATE_TIME_MODEL_FORMAT, useValue: ɵ4 },
                        { provide: DlDateAdapter, useClass: DlDateAdapterString }
                    ],
                },] }
    ];
    return DlDateTimeStringModule;
}());
export { DlDateTimeStringModule };
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4 };
